### ISSUE TITLE
Standard docstring formatting conventions proposal

### ISSUE THREAD

**Issue by jGaboardi** [pysal/pysal#1174]
## Standard docstring formatting conventions proposal

0. Based on [`numpy`](https://numpydoc.readthedocs.io/en/latest/format.html).
1. Summary starts directly after triple quotes on first line.
2. Empty line between all sections and elements within sections.
3. Valid sections are documented [here](https://numpydoc.readthedocs.io/en/latest/format.html#sections).
4. If the docstring is a single-line summary, put triple quotes on that line.
5. Variables should be returned whenever possible to aid in clarity and documentation.
6. Enforce proper spelling, grammar, and punctuation.
7. Elements within the Parameters and Returns section should be in the following format:
    ```
    <empty line>
    <Section>
    <empty line>
    <parameter/variable name><single space><:><single space><type>
    <four-space indent><description>
    <empty line>
    <Section>
    ```

### Basic example

```python
def main_func(arg1, arg2):
    """Perform the main function calculation.
    
    Parameters
    ----------
    
    arg1 : list
        Some values of data.
    
    arg2 : int
        Some multiplier.
    
    Returns
    -------
    
    result : list
        Some transformed values of data.
    
    Notes
    -----
    
    This is a sample equation within a docstring:
    
    .. math::
        
       \displaystyle E = mc^2
        
    See also
    --------
    
        some.other.Class
    
    Examples
    --------
    
    Call ``main_func()`` with these [ ] parameters.
    
    >>> main_func([3, 4], 2)
    [6, 8]
    
    """
    
    def helper_func(x):
        """Helper for calculating X."""
        new_x = x*arg2
        return new_x
    
    result = []
    for v in arg1:
        result.append(helper_func(v))
    
    return result
```

**Comment by jGaboardi** [comment]
The @pysal/devs team voted in favor of the standard in the [May meeting](https://hackmd.io/AnnMZhUARq6qnizL8ElWoA?both#adopt-docstring-standards-jGaboardi).

**Comment by jGaboardi** [comment]
@sjsrey @ljwolf I would like to create a [new project board](https://github.com/pysal/pysal/projects/new) for keeping track of progress, etc. on this. Where would it best reside: `pysal/pysal`, `pysal/libpysal`, some other repo, or no other repo?

**Comment by jGaboardi** [comment]
Also need to update:
 * [CONTRIBUTING.md](https://github.com/pysal/pysal/blob/master/.github/CONTRIBUTING.md)
 * [GitHub Standard Operating Procedures](https://github.com/pysal/pysal/wiki/GitHub-Standard-Operating-Procedures)
 * [Getting Started: For developers](http://pysal.org/getting_started#for-developers)

**Comment by sjsrey** [comment]
`pysal/pysal` makes sense to me

**Comment by slumnitz** [comment]
Thank you for initiating for bettering documentation @jGaboardi! 
I'd like to double check, the official NumPy [example](https://numpydoc.readthedocs.io/en/latest/example.html#example) does not contain lines between different `Parameters`,but only between section headers `Parameters`, `Returns`, ..., to keep the documentation compact. I would vouch for not including these extra lines, as documentation for functions with many parameters could get very lengthy/ easy to loose sight of parameter options. What do you think and could this be optional potentially?

**Comment by jGaboardi** [comment]
@slumnitz Yes, you are correct. My example does not exactly conform to `numpy`'s. Thanks for pointing this out. I will update the example and review the work I am currently doing in `libpysal`.

**Comment by jGaboardi** [comment]
An automated tool for `numpy`-style docstring reformatting to keep an eye on from [Matthias Bussonnier](https://github.com/Carreau): [see tweet](https://twitter.com/mbussonn/status/1281821312692834305?s=21)
